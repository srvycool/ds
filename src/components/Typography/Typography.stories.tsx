import React from 'react';
import { Typography } from './Typography';

export default {
  title: 'Typography',
  component: Typography,
};

export const All: React.FC = () => (
  <>
    <Typography variant="heading1">Heading 1</Typography>
    <Typography variant="heading2">Heading 2</Typography>
    <Typography variant="heading3">Heading 3</Typography>
    <Typography variant="subtitle">Subtitle</Typography>
    <br />
    <Typography variant="body1">Body 1</Typography>
    <br />
    <Typography variant="body2">Body 2</Typography>
    <br />
    <Typography variant="caption">Caption</Typography>
  </>
);
