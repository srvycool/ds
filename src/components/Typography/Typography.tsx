import React from 'react';
import styled, { StyledComponentProps } from 'styled-components';
import {
  space,
  color,
  variant,
  typography,
  TextColorProps,
  SpaceProps,
  TextAlignProps,
} from 'styled-system';
import { Theme } from '../../theme';

export type TypographyVariant =
  | 'heading1'
  | 'heading2'
  | 'heading3'
  | 'subtitle'
  | 'body1'
  | 'body2'
  | 'caption';

export type TypographyProps = {
  as?: keyof JSX.IntrinsicElements;
  variant: TypographyVariant;
} & TextColorProps<Theme> &
  SpaceProps<Theme> &
  TextAlignProps<Theme>;

const variantMapping: {
  [index in TypographyVariant]: keyof JSX.IntrinsicElements;
} = {
  heading1: 'h1',
  heading2: 'h2',
  heading3: 'h2',
  subtitle: 'span',
  body1: 'span',
  body2: 'span',
  caption: 'span',
};

const StyledTypography = styled.span<TypographyProps>`
  margin: 0;
  ${color}
  ${space}
  ${typography}
  ${variant({ scale: 'typography' })}
`;

export const Typography: React.FC<StyledComponentProps<
  'span',
  Theme,
  TypographyProps,
  never
>> = ({ as, variant, ...props }) => {
  return (
    <StyledTypography
      as={as || variantMapping[variant]}
      variant={variant}
      {...props}
    />
  );
};
