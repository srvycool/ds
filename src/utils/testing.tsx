import React from 'react';
import {
  render as tlrender,
  RenderOptions,
  RenderResult,
} from '@testing-library/react';
import { ThemeProvider } from '../components/ThemeProvider/ThemeProvider';

export function render(
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult {
  return tlrender(<ThemeProvider>{ui}</ThemeProvider>, options);
}
