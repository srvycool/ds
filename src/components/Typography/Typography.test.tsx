import React from 'react';
import { render } from '../../utils/testing';
import { Typography } from './Typography';

describe('Typography', () => {
  describe.each`
    variant
    ${'heading1'}
    ${'heading2'}
    ${'heading3'}
    ${'subtitle'}
    ${'body1'}
    ${'body2'}
    ${'caption'}
  `('$variant', ({ variant }) => {
    it('should match snapshot', () => {
      const { container } = render(
        <Typography variant={variant}>Text</Typography>
      );
      expect(container).toMatchSnapshot();
    });
  });
});
