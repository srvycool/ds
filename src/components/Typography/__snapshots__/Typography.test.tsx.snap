// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography body1 should match snapshot 1`] = `
.c0 {
  margin: 0;
  font-size: 16px;
  line-height: 1.2;
  font-family: Montserrat;
}

<div>
  <span
    class="c0"
  >
    Text
  </span>
</div>
`;

exports[`Typography body2 should match snapshot 1`] = `
.c0 {
  margin: 0;
  font-size: 12px;
  line-height: 1.2;
  font-family: Montserrat;
}

<div>
  <span
    class="c0"
  >
    Text
  </span>
</div>
`;

exports[`Typography caption should match snapshot 1`] = `
.c0 {
  margin: 0;
  font-size: 12px;
  line-height: 1.2;
  font-family: Montserrat;
  text-transform: uppercase;
}

<div>
  <span
    class="c0"
  >
    Text
  </span>
</div>
`;

exports[`Typography heading1 should match snapshot 1`] = `
.c0 {
  margin: 0;
  font-size: 24px;
  line-height: 1.2;
  font-family: Hammersmith One;
}

<div>
  <h1
    class="c0"
  >
    Text
  </h1>
</div>
`;

exports[`Typography heading2 should match snapshot 1`] = `
.c0 {
  margin: 0;
  font-size: 22px;
  line-height: 1.2;
  font-family: Hammersmith One;
}

<div>
  <h2
    class="c0"
  >
    Text
  </h2>
</div>
`;

exports[`Typography heading3 should match snapshot 1`] = `
.c0 {
  margin: 0;
  font-size: 20px;
  line-height: 1.2;
  font-family: Hammersmith One;
}

<div>
  <h2
    class="c0"
  >
    Text
  </h2>
</div>
`;

exports[`Typography subtitle should match snapshot 1`] = `
.c0 {
  margin: 0;
  font-size: 18px;
  line-height: 1.2;
  font-family: Montserrat;
}

<div>
  <span
    class="c0"
  >
    Text
  </span>
</div>
`;
